### PR TITLE
KAFKA-3133: Add putIfAbsent function to KeyValueStore

### PR DESCRIPTION
**Pull request by kichristensen**
@guozhangwang

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java [hunk 1]
```diff
@@ -79,6 +79,13 @@ public void put(K key, V value) {
     }
 
     @Override
+    public void putIfAbsent(K key, V value) {
+        if (get(key) == null) {
+            put(key, value);
+        }
+    }
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java:89]: I think it would be better to invoke `inner.putIfAbsent` so that we can benefit from a more efficient implementation where available.

**Comment by kichristensen** [comment]
@ijuma I've changed it to use `inner.putIfAbsent`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java [hunk 2]
```diff
@@ -79,6 +79,15 @@ public void put(K key, V value) {
     }
 
     @Override
+    public void putIfAbsent(K key, V value) {
+        if (get(key) == null) {
+            this.inner.putIfAbsent(key, value);
+            changeLogger.add(key);
+            changeLogger.maybeLogChange(this.getter);
+        }
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java:87]: You don't need to call `get` in this case. Use the return value of `pufIfAbsent` instead.

Comment (kichristensen) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java:87]: But then how will I know if the key should be added to the change logger?

**Comment by kichristensen** [comment]
Also I have no idea why this build failed?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java (hunk 2 above)
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java:87]: I noticed now that `putIfAbsent` doesn't return anything. The typical signature is usually like:

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#putIfAbsent-K-V-

Comment (kichristensen) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java:87]: Yes, I would probably be best to follow use the same signature here, I'll change this

**Comment by kichristensen** [comment]
@ijuma The signature has now been changed to follow the normal behaviour

**Comment by guozhangwang** [comment]
@ijuma Do you want to take another look?

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java [hunk 3]
```diff
@@ -51,6 +51,16 @@
     abstract public void put(K key, V value);
 
     /**
+     * Update the value associated with this key, unless a value
+     * already is associated with the key
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: Nitpick: `unless a value is already associated` reads better, I think. We need to document the return value since we have now added that. Also, we should make it clear that this may or may not be atomic (as people will expect it to be atomic given `ConcurrentMap.putIfAbsent`).

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java [hunk 4]
```diff
@@ -192,4 +192,41 @@ public void testRestoreWithDefaultSerdes() {
...
+        KeyValueStoreTestDriver<Integer, String> driver = KeyValueStoreTestDriver.create(Integer.class, String.class);
+        KeyValueStore<Integer, String> store = createKeyValueStore(driver.context(), Integer.class, String.class, true);
+        try {
+
+            // Verify that the store reads and writes correctly ...
+            store.putIfAbsent(0, "zero");
+            store.putIfAbsent(1, "one");
+            store.putIfAbsent(2, "two");
+            store.putIfAbsent(4, "four");
+            store.putIfAbsent(4, "unexpected value");
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractKeyValueStoreTest.java]: We should also assert that the return value is correct.

**Comment by ijuma** [comment]
@kichristensen Thanks, I left a few minor comments, it looks good otherwise. Having said that, I am not very familiar with the streams code, so @guozhangwang please double-check.

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java (hunk 3 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: Since a state store is only accessed by a single thread, it should be atomic by itself. But we can make this statement in the comment.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreSupplier.java [hunk 5]
```diff
@@ -117,6 +117,15 @@ public void put(K key, V value) {
         }
 
         @Override
+        public V putIfAbsent(K key, V value) {
+            V originalValue = get(key);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreSupplier.java:121]: I see this can be simpler if `put` call itself can also return the old value. But this would better be in a separate PR.

**Comment by guozhangwang** [comment]
@tomdearman LGTM overall, could you address @ijuma 's comments before merging?

**Comment by tomdearman** [comment]
@guazhangwang, I haven't worked on this I think you intended it for someone
else.

**Comment by guozhangwang** [comment]
My bad, I meant to @kichristensen

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java (hunk 3 above)
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: @guozhangwang The reason why I said the above is that `putIfAbsent` is a method that exists in `ConcurrentMap` and atomicity is guaranteed when it is called by multiple threads. We can write it differently (eg say that this is meant to be called from a single thread only), but I would not claim atomicity in that case (it's a bit of a meaningless statement in my opinion).

**Comment by kichristensen** [comment]
@ijuma @guozhangwang I've addressed the comments. Please review again

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java (hunk 3 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java]: I see, that makes sense.

**Comment by guozhangwang** [comment]
LGTM. Jenkins failures are due to transient known errors.

Merged to trunk.
